typing/typecore.ml

```diff
@@ -3059,23 +3101,31 @@ and type_expect_
             Some exp
       in
       let ty_record, expected_type =
-        let get_path ty =
-          try
-            let (p0, p,_) = extract_concrete_record env ty in
-            let principal =
-              (repr ty).level = generic_level || not !Clflags.principal
+        let get_path loc expectation ty =
+          match extract_concrete_record env ty with
+          | Record_type (p0, p, _) ->
+              let principal =
+                (repr ty).level = generic_level || not !Clflags.principal
+              in
+              Some (p0, p, principal)
+          | Maybe_a_record_type -> None
+          | Not_a_record_type ->
+            let error =
+              if expectation then
+                Wrong_expected_kind(Record, Expression explanation, ty)
+              else
+                Expr_not_a_record_type ty
             in
-            Some (p0, p, principal)
-          with Not_found -> None
+            raise (Error (loc, env, error))
         in
-        let opath = get_path ty_expected in
+        let opath = get_path loc true ty_expected in
```

Comment:
for me an un-labelled boolean argument is a code smell. You could use `get_path loc ~expectation:true ty_expected`.

But maybe there is an even clearer way to distinguish between the two situations here? On a first read I'm not sure I understand what `Expr_not_a_record_type` means compared to `Wrong_expected_kind`, and why we don't have a symmetric case for (non-polymorphic) variant constructors.

Reply:
Could this suggest that `type_forcing_context` should be extended with a `Record_project` case or something?

Reply:
This code was getting hard to follow so I refactored it. Should be clearer now.

The difference between the two errors is that one is about using a constructor with an inappropriate expected type, and the other is about using a projection on an expression of the wrong type. You only get the second kind of error for records since there are no projections for variants.